pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+                Some(ptr) => match ImportedHashes::<T>::get(ptr) {
+                    Some(key) => ImportedBlocks::<T>::get(key),
+                    None => None,
+                },
+                None => None,
+            }
+        }
+    }
+
+    /// If true, stop the world
+    #[pallet::storage]
+    #[pallet::getter(fn is_halted)]
+    pub type IsHalted<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn get_options)]
+    pub type LightClientOptions<T: Config> = StorageValue<_, LightClientOptionsStorage, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        // TODO : benchmark & adjust weights
+        #[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+        pub fn initialize_client(
+            origin: OriginFor<T>,
+            options: LightClientOptionsStorage,
+            initial_block: LightBlockStorage,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            log::debug!(target: "runtime::tendermint-lc", "Initializing Light Client light:\n options: {:?}\n initial block {:#?}", &options, &initial_block);
+
+            // ensure client is not already initialized
+            let can_initialize = !<LastImportedBlockHash<T>>::exists();
+            ensure!(can_initialize, <Error<T>>::AlreadyInitialized);
+
+            match initial_block.signed_header.commit.block_id.hash {
+                TendermintHashStorage::Some(hash) => {
+                    <LightClientOptions<T>>::put(options);
+                    <ImportedHashesPointer<T>>::put(0);
+                    // update block storage
+                    insert_light_block::<T>(hash, initial_block);
+
+                    // update status
+                    <IsHalted<T>>::put(false);
+                    log::info!(target: "runtime::tendermint-lc", "Light client initialized");
+                    Self::deposit_event(Event::LightClientInitialized);
+                    Ok(())
+                }
+                TendermintHashStorage::None => {
+                    log::warn!(target: "runtime::tendermint-lc", "Rejecting invalid initial light block; empty hash");
+                    fail!(<Error<T>>::InvalidInitialBlock)
+                }
+            }
+        }
+
+        // TODO : benchmark & adjust weights
+        /// Verify a block header against a known state.        
+        #[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+        pub fn update_client(
+            origin: OriginFor<T>,
+            untrusted_block: LightBlockStorage,
+        ) -> DispatchResult {
+            ensure_not_halted::<T>()?;
+            let who = ensure_signed(origin)?;
+            let hash = untrusted_block.signed_header.commit.block_id.hash;
+
+            log::debug!(target: "runtime::tendermint-lc", "Verifying light block {:#?}", &hash);
```

Comment:
Let's log height as well - like previously, Tendermint uses heights more than hashes. It will be easier to correlate things if we have both informations.

Reply:
done